pkg/reconciler/cache/replication/replication_reconcile_unstructured.go

```diff
@@ -0,0 +1,204 @@
+	}
+	specChanged, err := ensureUnstructuredSpec(unstructuredCacheObject, unstructuredLocalObject)
+	if err != nil {
+		return err
+	}
+	statusChanged, err := ensureUnstructuredStatus(unstructuredCacheObject, unstructuredLocalObject)
+	if err != nil {
+		return err
+	}
+	if !specChanged && !metaChanged && !statusChanged {
+		return nil
+	}
+
+	if specChanged || metaChanged || statusChanged {
+		_, err := c.dynamicCacheClient.Cluster(cluster).Resource(*gvr).Namespace(unstructuredCacheObject.GetNamespace()).Update(ctx, unstructuredCacheObject, metav1.UpdateOptions{})
+		return err
+	}
+	return nil
+}
+
+func (c *controller) handleUnstructuredObjectDeletion(ctx context.Context, cluster logicalcluster.Name, gvr *schema.GroupVersionResource, unstructuredCacheObject *unstructured.Unstructured) error {
+	if unstructuredCacheObject == nil {
+		return nil // the cached object already removed
+	}
+	if unstructuredCacheObject.GetDeletionTimestamp() == nil {
+		return c.dynamicCacheClient.Cluster(cluster).Resource(*gvr).Delete(ctx, unstructuredCacheObject.GetName(), metav1.DeleteOptions{})
+	}
+	return nil
+}
+
+// ensureUnstructuredMeta changes unstructuredCacheObject's metadata to match unstructuredLocalObject's metadata.
+// returns true when the unstructuredCacheObject was updated.
+func ensureUnstructuredMeta(unstructuredCacheObject *unstructured.Unstructured, unstructuredLocalObject *unstructured.Unstructured) (bool, error) {
+	var shardName string
+	cacheObjectAnnotations := unstructuredCacheObject.GetAnnotations()
+	if cacheObjectAnnotations != nil {
+		shardName = cacheObjectAnnotations[genericrequest.AnnotationKey]
+		delete(cacheObjectAnnotations, genericrequest.AnnotationKey)
+	}
+
+	localObjectAnnotations := unstructuredLocalObject.GetAnnotations()
+	localObjectLabels := unstructuredLocalObject.GetLabels()
+	objectMetaChanged := !equality.Semantic.DeepEqual(unstructuredCacheObject.GetLabels(), localObjectLabels) || !equality.Semantic.DeepEqual(cacheObjectAnnotations, localObjectAnnotations)
+	if !objectMetaChanged {
+		return false, nil
+	}
+
+	cacheObjectNewAnnotations := map[string]string{}
+	for k, v := range localObjectAnnotations {
+		cacheObjectNewAnnotations[k] = v
+	}
+	if len(shardName) > 0 {
+		cacheObjectNewAnnotations[genericrequest.AnnotationKey] = shardName
+	}
+	unstructuredCacheObject.SetAnnotations(cacheObjectNewAnnotations)
+
+	cacheObjectNewLabels := map[string]string{}
+	for k, v := range localObjectLabels {
+		cacheObjectNewLabels[k] = v
+	}
+	unstructuredCacheObject.SetLabels(cacheObjectNewLabels)
+
+	return true, nil
+}
+
+// ensureUnstructuredSpec changes unstructuredCacheObject's spec to match unstructuredLocalObject's spec.
+// returns true when the unstructuredCacheObject was updated.
+//
+// note that the spec field is not required, in that case this function is no-op
+func ensureUnstructuredSpec(unstructuredCacheObject *unstructured.Unstructured, unstructuredLocalObject *unstructured.Unstructured) (bool, error) {
```

Comment:
am confused that we even distinguish between spec, status, ... Why not just update everything?